tutorials/application_design/py/subscriber.py

```diff
@@ -30,86 +27,122 @@ class DashboardItem:
     current_destination: typing.Optional[Coord] = None
-    reached_destinations: typing.List[Coord] = dataclasses.field(
-        default_factory=list
-    )
 
 
 class SubscriberDashboard:
     def __init__(
         self,
-        metrics_reader: "dds.DataReader",
-        transit_reader: "dds.DataReader",
-    ):
+        metrics_reader: dds.DataReader,
+        transit_reader: dds.DataReader,
+    ) -> None:
         self._metrics_reader = metrics_reader
         self._transit_reader = transit_reader
-        self._dashboard_data: typing.Dict[
-            dds.InstanceHandle, DashboardItem
-        ] = dict()
 
     def __repr__(self):
-        return f"Dashboard({self._metrics_reader=}, {self._transit_reader=}, {self._dashboard_data=})"
+        return f"Dashboard({self._metrics_reader=}, {self._transit_reader=})"
+
+    def run(self):
+
+        # Create a new handler for newly-received data.
+        # This handler will be called as data comes by, by dispatching
+        # the WaitSet that has attached this ReadCondition.
+        def new_position_handler(_):
+            print(self._create_new_position_string(new_position_condition))
+
+        new_position_condition = dds.ReadCondition(
+            self._transit_reader, dds.DataState.new_data, new_position_handler
+        )
+
+        # Create a new handler for printing the dashboard.
+        # This handler will be called whenever the GuardCondition is triggered,
+        # which we're periodically doing on a background thread.
+        def dashboard_handler(_):
+            print(self._create_dashboard_string())
+
+        dashboard_condition = dds.GuardCondition()
+        dashboard_condition.set_handler(dashboard_handler)
+
+        # Create a background thread that will trigger the dashboard condition.
+        def display_handler():
+            dashboard_condition.trigger_value = True
+
+        display_thread = threading.Thread(target=display_handler)
+
+        # Create a WaitSet and attach the conditions to it.
+        waitset = dds.WaitSet()
+        waitset.attach_condition(new_position_condition)
+        waitset.attach_condition(dashboard_condition)
+
+        # Start the thread now that the WaitSet has been created.
+        display_thread.start()
+
+        try:
+            while True:
+                waitset.dispatch()
+                dashboard_condition.trigger_value = False
+        except KeyboardInterrupt:
+            pass
+
+    def _create_new_position_string(self, condition):
+        string = ""
+        for sample, info in (
+            self._transit_reader.select().condition(condition).read()
+        ):
+            if not info.valid:
+                continue
 
-    async def run(self):
-        await asyncio.gather(
-            self._display_app(),
-            self._metrics_app(),
-            self._transit_app(),
+            string += f"[INFO] Vehicle {sample.vehicle_vin}"
+            if sample.current_route and len(sample.current_route):
```

Comment:
```
if sample.current_route:
```
The and is always true.

Reply:
I don't think this is accurate as `if sample.current_route` will evaluate to `sample.current_route.__bool__` which tests for whether if the optional is present or not. Once we've checked that, we need to check that the sequence is not empty.

Reply:
Both None and an empty list/string test to `False`:

```
>>> bool([])
False
>>> bool(None)
False
>>> bool([1])
True
```